### DISCUSSION TITLE
Can a rule be specified multiple times without overriding itself?

### DISCUSSION THREAD

**Discussion by fregante** [eslint/eslint#15011] [Q&A]
Some rules are very flexible, like https://eslint.org/docs/rules/no-restricted-imports

This means I can just use the rule to block anything I want:

```json
"no-restricted-imports": ["error", {
    "paths": ["import1", "import2"],
    "patterns": ["import1/private/*", "import2/*", "!import2/good"]
}]
```

What can I do though when I want to enforce some rules as `error` and some as `warn`? impossible 👇 

```json
"no-restricted-imports": ["error", { "paths": ["never-import"] }],
"no-restricted-imports": ["warn", { "paths": ["just-a-warning"] }]
```

Also the same happens for overrides. What if I want a global setting _and_ an override for some other files? Impossible, 👇 the override completely replaces the first definition.

```json
rules: {
  "no-restricted-imports": ["error", { "paths": ["no-lodash"] }]
},
overrides: [{
  "files": "*.stories.tsx",
  "rules": {
    "no-restricted-imports": ["error", { "paths": ["also-react-native"] }],
  }
}]
```

I think the only way to do this would be, if it existed, to make the same rule available under a different name so that it can be specified multiple times, example:

```json
"no-restricted-imports": ["error", { "paths": ["never-import"] }],
"alias-a/no-restricted-imports": ["warn", { "paths": ["just-a-warning"] }],
"alias-b/no-restricted-imports": ["warn", { "paths": ["jQuery"] }]
```

👆 not real 😕

**Comment by nzakas** [comment]
There isn’t a way to do this currently. With the new config system, it will be possible to do this for plug-in rules, but not for core rules.

**Reply by meandmax** [reply]
Hey @nzakas 

do you know when and how we would be able to do this? :)

**Reply by nzakas** [reply]
Follow https://github.com/eslint/eslint/issues/13481

**Comment by audunolsen** [comment]
I'm stumped on the exact same issue as the author of this discusson, @fregante. With the introduction of the flat config style, is there now a way to achieve this? @nzakas

**Reply by fregante** [reply]
I haven't had time/interest to explore this, but I just asked Claude and it proposed this.

1. it declares a new inline `myPlugin` plugin
2. it re-exports a core rule from eslint
3. it configures this new re-exported rule under a new plugin name

Untested, worth a try

```js
// eslint.config.js
import eslint from '@eslint/js';

export default [
  {
    files: ['**/*.{js,jsx,mjs,cjs,ts,tsx}'],
    ignores: ['**/node_modules/**', 'dist/**'],
    plugins: {
      myPlugin: {
        rules: {
          // Re-export the native no-restricted-imports rule
          'no-restricted-imports': eslint.configs.recommended.rules['no-restricted-imports'],
        },
      },
    },
    rules: {
      // Use the re-exported rule
      'myPlugin/no-restricted-imports': ['error', {
        paths: [
          {
            name: 'lodash',
            message: 'Please use individual lodash functions instead.',
          },
        ],
        patterns: [
          {
            group: ['lodash/*'],
            message: 'Please use individual lodash functions instead.',
          },
        ],
      }],
    },
  },
];
```

**Reply by aarongoldenthal** [reply]
Just to see, I ran it (with `eslint.configs.all` instead of `eslint.configs.recommended`, since that rule is not in `recommended`), and with several variations and they all failed. The core issue I see is that the rules themselves are not exported, e.g. the value of `eslint.configs.recommended.rules['no-restricted-imports']` is `"error"`, so it can't be defined as a different rule, and you get `Error while processing options validation schema of rule 'myPlugin/no-restricted-imports'`.

With plugins it's straightforward since you can give an arbitrary name for the plugin, for example:

```js
const sonarjsPlugin = require('eslint-plugin-sonarjs');

module.exports = {
    files: ['**/*.{js,mjs,cjs}'],
    plugins: { sonarjs: sonarjsPlugin, sonarjs2: sonarjsPlugin },
    rules: {
        'sonarjs/cognitive-complexity': [ 'warn', 5 ],
        'sonarjs2/cognitive-complexity': [ 'error', 10 ]
    }
};
```

**Reply by fregante** [reply]
Indeed the actual rule actually lives at `eslint/lib/rules/no-restricted-imports.js`, which I think is blocked by the `exports` map. I see a `/rules` export so probably it's:

```js
import rules from 'eslint/rules'
```

But in reality it's a type-only export, so yeah. I opened an issue for visibility:

- https://github.com/eslint/eslint/issues/19013

**Comment by aarongoldenthal** [comment]
@nzakas [Here](https://github.com/eslint/eslint/issues/15620#issuecomment-1048388514) you mentioned the possibility of exposing the core eslint rules so this could be accomplished for those as well as for plugins with flat config (where it does work). I don't see it in the exports, but also didn't see it covered specifically in another issue. I'm just trying to confirm it wasn't ruled out somewhere before opening a feature request (specifically thinking of cases like `complexity` - `warn` at 5, but `error` at 10).
